src/appleseed/renderer/modeling/aov/cryptomatte.cpp

```diff
@@ -0,0 +1,736 @@
+            footprint = AABB2i::intersect(footprint, crop_window);
+
+            // Bail out if the point does not fall inside the crop window.
+            // Only check the x coordinate; the y coordinate is checked in the loop below.
+            if (footprint.min.x > footprint.max.x)
+                return;
+
+            for (int ry = footprint.min.y; ry <= footprint.max.y; ++ry)
+            {
+                for (int rx = footprint.min.x; rx <= footprint.max.x; ++rx)
+                {
+                    const float weight = m_renderer_filter->evaluate(rx - dx, ry - dy);
+                    m_pixel_samples[(m_tile_origin_y + ry) * m_frame_width + (m_tile_origin_x + rx)][m3hash] += weight; // possible race conditions writing to pixels on the tile edges
+                }
+            }
+            return;
+        }
+    };
+
+    const char* CryptomatteAOVModel = "cryptomatte_aov";
+}
+
+
+//
+// Cryptomatte AOV class implementation.
+//
+
+struct CryptomatteAOV::Impl
+{
+    vector<WeightMap>                   m_pixel_samples;
+    map<pair<int, int>, NameMap>        m_tile_name_map;
+    unique_ptr<Image>                   m_image;
+    int                                 m_num_layers;
+    CryptomatteAOV::CryptomatteType     m_layer_type;
+};
+
+CryptomatteAOV::CryptomatteAOV(
+    int num_layers,
+    CryptomatteType layer_type)
+  : AOV("cryptomatte", ParamArray())
+  , impl(new Impl())
+{
+    impl->m_num_layers = num_layers;
+    impl->m_layer_type = layer_type;
+}
+
+CryptomatteAOV::~CryptomatteAOV()
+{
+    delete impl;
+}
+
+const char* CryptomatteAOV::get_model() const
+{
+    return CryptomatteAOVModel;
+}
+
+size_t CryptomatteAOV::get_channel_count() const
+{
+    return 0;
+}
+
+const char** CryptomatteAOV::get_channel_names() const
+{
+    return nullptr;
+}
+
+bool CryptomatteAOV::has_color_data() const
+{
+    return false;
+}
+
+void CryptomatteAOV::create_image(
+    const size_t    canvas_width,
+    const size_t    canvas_height,
+    const size_t    tile_width,
+    const size_t    tile_height,
+    ImageStack&     aov_images)
+{
+    const int w = static_cast<int>(canvas_width);
+    const int h = static_cast<int>(canvas_height);
+
+    // Create underlying images.
+    const int num_channels = (impl->m_num_layers * 2) + 3;
+    impl->m_image.reset(
+        new Image(
+            canvas_width,
+            canvas_height,
+            tile_width,
+            tile_height,
+            num_channels,
+            PixelFormatFloat));
+    impl->m_pixel_samples.resize(w * h);
+    clear_image();
+}
+
+void CryptomatteAOV::clear_image()
+{
+    const int num_channels = (impl->m_num_layers * 2) + 3;
+    for (size_t ry = 0, e = impl->m_image->properties().m_canvas_height; ry < e; ++ry)
+    {
+        for (size_t rx = 0, e = impl->m_image->properties().m_canvas_width; rx < e; ++rx)
+        {
+                vector<float> pixel_values;
+                for (size_t i = 0; i < num_channels; ++i)
```

Comment:
Move this out of the loop. Use the vector constructor that creates N elements with a value.